storage/innobase/arch/arch0arch.cc

```diff
@@ -0,0 +1,640 @@
+  ut_ad(m_archiver_active);
+
+  struct timespec wait_time;
+  while(!m_signalled)
+  {
+    set_timespec(wait_time, 1);
+    mysql_cond_timedwait(&m_cond, &m_mutex, &wait_time);
+  }
+  m_signalled= false;
+  mysql_mutex_unlock(&m_mutex);
+}
+
+void Arch_Sys::archiver_stopped()
+{
+  mysql_mutex_lock(&m_mutex);
+  m_archiver_active= false;
+  mysql_mutex_unlock(&m_mutex);
+}
+
+void Arch_Sys::remove_file(const char *file_path, const char *file_name)
+{
+  char path[MAX_ARCH_PAGE_FILE_NAME_LEN];
+
+  static_assert(MAX_ARCH_LOG_FILE_NAME_LEN <= MAX_ARCH_PAGE_FILE_NAME_LEN);
+  ut_ad(strlen(file_path) + 1 + strlen(file_name) <
+        MAX_ARCH_PAGE_FILE_NAME_LEN);
+
+  /* Remove only LOG and PAGE archival files. */
+  if (0 != strncmp(file_name, ARCH_LOG_FILE, strlen(ARCH_LOG_FILE)) &&
+      0 != strncmp(file_name, ARCH_PAGE_FILE, strlen(ARCH_PAGE_FILE)) &&
+      0 != strncmp(file_name, ARCH_PAGE_GROUP_DURABLE_FILE_NAME,
+                   strlen(ARCH_PAGE_GROUP_DURABLE_FILE_NAME)))
+    return;
+
+  snprintf(path, sizeof(path), "%s%c%s", file_path, OS_PATH_SEPARATOR,
+           file_name);
+
+#ifdef UNIV_DEBUG
+  os_file_type_t type;
+  bool exists;
+
+  os_file_status(path, &exists, &type);
+  ut_ad(exists);
+  ut_ad(type == OS_FILE_TYPE_FILE);
+#endif /* UNIV_DEBUG */
+
+  os_file_delete(innodb_arch_file_key, path);
+}
+
+void Arch_Sys::remove_dir(const char *dir_path, const char *dir_name)
+{
+  char path[MAX_ARCH_DIR_NAME_LEN];
+
+  static_assert(sizeof(ARCH_LOG_DIR) <= sizeof(ARCH_PAGE_DIR));
+  ut_ad(strlen(dir_path) + 1 + strlen(dir_name) + 1 < sizeof(path));
+
+  /* Remove only LOG and PAGE archival directories. */
+  if (0 != strncmp(dir_name, ARCH_LOG_DIR, strlen(ARCH_LOG_DIR)) &&
+      0 != strncmp(dir_name, ARCH_PAGE_DIR, strlen(ARCH_PAGE_DIR)))
+    return;
+
+  snprintf(path, sizeof(path), "%s%c%s", dir_path, OS_PATH_SEPARATOR, dir_name);
+
+#ifdef UNIV_DEBUG
+  os_file_type_t type;
+  bool exists;
+
+  os_file_status(path, &exists, &type);
+  ut_ad(exists);
+  ut_ad(type == OS_FILE_TYPE_DIR);
+#endif /* UNIV_DEBUG */
+
+  os_file_scan_directory(path, Arch_Sys::remove_file, true);
+}
+
+/** Initialize Page and Log archiver system. */
+Arch_Sys::Arch_Sys()
+{
+  mysql_mutex_init(0, &m_mutex, nullptr);
+  mysql_cond_init(0, &m_cond, nullptr);
+
+  m_signalled= false;
+  m_archiver_active= false;
+
+  if (srv_read_only_mode)
+    m_page_sys.set_read_only_mode();
+  else
+    m_page_sys.recover();
+}
+
+/** Free Page and Log archiver system */
+Arch_Sys::~Arch_Sys()
+{
+  mysql_cond_destroy(&m_cond);
+  mysql_mutex_destroy(&m_mutex);
+}
+
+dberr_t Arch_Group::write_to_file(Arch_File_Ctx *from_file, byte *from_buffer,
+                                  uint length, bool partial_write,
+                                  bool do_persist)
+{
+  dberr_t err= DB_SUCCESS;
+  uint write_size;
+
+  if (m_file_ctx.is_closed())
+  {
+    /* First file in the archive group. */
+    ut_ad(m_file_ctx.get_count() == 0);
```

Comment:
There is only one `ib_logfile0` in MariaDB, and this is about log archiving. What is an "archive group"? The parameter `innodb_log_files_in_group` was removed from MariaDB.

Also, if this subsystem is only about log archiving, why don’t we just append to the archived log file in `log_t::write_buf()`, similar to how we maintain an extra copy `re_write_buf` for `ib_logfile101` during log resizing?